src/psyclone/psyad/tl2ad.py

```diff
@@ -98,9 +155,407 @@ def generate_adjoint(tl_psyir):
+            # use `new_symbol` to ensure there are no name clashes.
+            new_sym = symbol_table.new_symbol(arg.name, symbol_type=DataSymbol,
+                                              datatype=arg.datatype)
+        else:
+            # Since a Symbol's shape is immutable, we have to create a new
+            # symbol in case the argument is of assumed size.
+            new_shape = []
+            for dim in arg.datatype.shape:
+                if isinstance(dim, ArrayType.Extent):
+                    # This dimension is assumed size so we have to give it
+                    # an explicit size in the test program.
+                    new_shape.append(Reference(dim_size_sym))
+                elif isinstance(dim, Reference):
+                    if dim.symbol in new_dim_args_map:
+                        new_shape.append(
+                            Reference(new_dim_args_map[dim.symbol]))
+                    else:
+                        raise NotImplementedError(
+                            "Found argument '{0}' to kernel '{1}' which has a "
+                            "reference to '{2}' in its shape. However, '{2}' "
+                            "is not passed as an argument. This is not "
+                            "supported.".format(
+                                arg.name, tl_kernel.name, dim.symbol.name))
+                elif isinstance(dim, Literal):
+                    new_shape.append(dim.copy())
+                else:
+                    raise InternalError(
+                        "Argument '{0}' to kernel '{1}' contains a '{2}' in "
+                        "its shape definition but expected an ArrayType."
+                        "Extent, a Reference or a Literal".format(
+                            arg.name, tl_kernel.name, type(dim).__name__))
+            new_sym = symbol_table.new_symbol(arg.name, symbol_type=DataSymbol,
+                                              datatype=ArrayType(arg.datatype,
+                                                                 new_shape))
+        new_arg_list.append(new_sym)
+        # Create variables to hold a copy of the inputs
+        input_sym = symbol_table.new_symbol(new_sym.name+"_input",
+                                            symbol_type=type(new_sym),
+                                            datatype=new_sym.datatype)
+        inputs.append(new_sym)
+        input_copies.append(input_sym)
+
+    statements = []
+    # Initialise those variables and keep a copy of them.
+    # TODO #1247 we need to add comments to the generated code!
+    for sym, sym_copy in zip(inputs, input_copies):
+        # The PSyIR doesn't support the random_number Fortran intrinsic so we
+        # create a CodeBlock for it. Happily, the intrinsic will initialise
+        # all elements of an array passed to it so we don't have to take any
+        # special action.
+        ptree = Fortran2003.Call_Stmt(
```

Comment:
I obviously don't like this as we are now Fortran-specific. Perhaps add an issue that will look to address this.

Reply:
Yes. Created #1345 and referenced it in the code.